ReactiveCocoa/Swift/Signal.swift

```diff
@@ -299,24 +299,168 @@ extension SignalType {
+	func append(value: Value) {
 		values.append(value)
-		return self
+	}
+
+	/// Check if there are any items remaining.
+	///
+	/// - Note: This method also checks if there weren't collected any values 
+	/// and, in that case, it means an empty array should be sent as the result
+	/// of collect.
+	var isEmpty: Bool {
+		/// We use capacity being zero to determine if we haven't collected any 
+		/// value since we're keeping the capacity of the array to avoid 
+		/// unnecessary and expensive allocations). This also guarantees 
+		/// retro-compatibility around the original `collect()` operator.
+		return values.isEmpty && values.capacity > 0
+	}
+
+	/// Removes all values previously collected if any.
+	func flush() {
+		// Minor optimization to avoid consecutive allocations. Can
+		// be useful for sequences of regular or similar size and to
+		// track if any value was ever collected.
+		values.removeAll(keepCapacity: true)
 	}
 }
 
+
 extension SignalType {
-	/// Returns a signal that will yield an array of values when `self` completes.
+
+	/// Returns a signal that will yield an array of values until it reaches a 
+	/// certain count.
+	///
+	/// When the count is reached the array is sent and the signal starts over 
+	/// yielding a new array of values.
+	///
+	/// - Precondition: `count` should be greater than zero.
+	///
 	@warn_unused_result(message="Did you forget to call `observe` on the signal?")
-	public func collect() -> Signal<[Value], Error> {
-		return self
-			.reduce(CollectState()) { $0.append($1) }
-			.map { $0.values }
+	public func collect(count count: Int) -> Signal<[Value], Error> {
+		precondition(count > 0)
+		return collect { values in values.count == count }
+	}
+
+	/// Returns a signal that will yield an array of values based on a predicate
+	/// which matches the values collected.
+	///
+	/// - parameter predicate: Predicate to match when values should be sent
+	/// (returning `true`) or alternatively when they should be collected (where
+	/// it should return `false`). The predicate receives the `values` already 
+	/// collected which will already include the next value. By default, uses a
+	/// predicate that collects all the values until `self` completes.
+	///
+	/// #### Example
+	///
+	///     let (signal, observer) = Signal<Int, NoError>.pipe()
+	///
+	///     signal
+	///         .collect { values in values.reduce(0, combine: +) <= 8 }
+	///         .observeNext { print($0) }
+	///
+	///     observer.sendNext(1)
+	///     observer.sendNext(3)
+	///     observer.sendNext(4)
+	///     observer.sendNext(7)
+	///     observer.sendNext(1)
+	///     observer.sendNext(5)
+	///     observer.sendNext(6)
+	///     observer.sendCompleted()
+	///
+	///     // Output:
+	///     // [1, 3, 4]
+	///     // [7, 1]
+	///     // [5, 6]
+	///
+	@warn_unused_result(message="Did you forget to call `observe` on the signal?")
+	public func collect(predicate: (values: [Value]) -> Bool = { _ in false }) -> Signal<[Value], Error> {
```

Comment:
I think it might be better to use a separate overloaded method for `collect()`, rather than a default value here.